src/main.rs

```diff
@@ -295,113 +299,95 @@ fn main() {
+            future::poll_fn(move || signals.lock().expect("Not poisoned").poll())
         };
+        signals
+            .lock()
+            .expect("Not poisoned")
+            .add_as(
+                graceful_crash.map_err::<String, _>(|()| unreachable!()),
+                SignalEvent::Shutdown,
+            )
+            .add_as(
+                config_watcher.map_err::<String, _>(|()| unreachable!()),
+                SignalEvent::ReloadConfig,
+            );
 
-        if signal == SIGINT || signal == SIGTERM {
-            info!("Shutting down.");
-            let shutdown = topology.stop();
-            metrics_trigger.cancel();
+        #[cfg(unix)]
+        {
+            signals
+                .lock()
+                .expect("Not poisoned")
+                .add_as(Signal::new(SIGINT).flatten_stream(), SignalEvent::Shutdown)
+                .add_as(Signal::new(SIGTERM).flatten_stream(), SignalEvent::Shutdown)
+                .add_as(Signal::new(SIGQUIT).flatten_stream(), SignalEvent::Quit)
+                .add_as(
+                    Signal::new(SIGHUP).flatten_stream(),
+                    SignalEvent::ReloadConfig,
+                );
+        }
+        #[cfg(windows)]
+        {
+            signals.lock().expect("Not poisoned").add_as(
+                tokio_signal::ctrl_c().flatten_stream(),
+                SignalEvent::Shutdown,
+            );
+        }
 
-            match rt.block_on(shutdown.select2(signals.into_future())) {
-                Ok(Either::A(_)) => { /* Graceful shutdown finished */ }
-                Ok(Either::B(_)) => {
-                    info!("Shutting down immediately.");
-                    // Dropping the shutdown future will immediately shut the server down
+        // Process signals in running state
+        loop {
+            let signals = signals.clone();
+            match rt.block_on(signal_future()) {
+                Ok(Some(SignalEvent::ReloadConfig)) => reload_config(&opts, &mut topology, &mut rt),
+                Ok(Some(SignalEvent::Shutdown)) => {
+                    info!("Shutting down.");
+                    let shutdown = topology.stop();
+                    metrics_trigger.cancel();
+
+                    signals.lock().expect("Not poisoned").add_as(
+                        shutdown
+                            .into_stream()
+                            .map_err::<String, _>(|()| unreachable!()),
+                        SignalEvent::Stopped,
+                    );
+
+                    // Process signals in shutdown state
+                    loop {
+                        match rt.block_on(signal_future()) {
+                            Ok(Some(SignalEvent::ReloadConfig)) => (),
+                            Ok(Some(SignalEvent::Shutdown)) | Ok(Some(SignalEvent::Quit)) => {
+                                info!("Shutting down immediately.");
+                                // Dropping the shutdown future will immediately shut the server down
```

Comment:
This to me adds some implicit dependency on these futures blocking stuff in other places, would it be better here to have a named scope on the outter loop and break out of that at the break below?

Reply:
It would.